discojs/discojs-core/src/models/gpt/evaluate.ts

```diff
@@ -1,50 +1,58 @@
 import * as tf from '@tensorflow/tfjs'
 
+interface DataPoint extends tf.TensorContainerObject {
+  xs: tf.Tensor2D,
+  ys: tf.Tensor3D,
+}
+
 export default async function evaluate (
   model: tf.LayersModel,
-  dataset: tf.data.Dataset<{ xs: tf.Tensor, ys: tf.Tensor }>
+  dataset: tf.data.Dataset<{ xs: tf.Tensor, ys: tf.Tensor }>,
+  maxEvalBatches: number
 ): Promise<Record<'acc' | 'val_acc' | 'val_loss' | 'val_perplexity', number>> {
   let datasetSize = 0
   let totalLoss = 0
   const acc: [number, number] = [0, 0]
 
-  await dataset.map(({ xs, ys }) => {
+  let iteration = 1
+  const iterator = await dataset.iterator()
+  let continueIterating = true
+  while (continueIterating) {
+    const next = await iterator.next()
+
+    const { xs, ys } = next.value as DataPoint
     const logits = model.apply(xs)
     if (Array.isArray(logits)) {
-      throw new Error('model outputed many tensor')
+      throw new Error('model output too many tensor')
     }
     if (logits instanceof tf.SymbolicTensor) {
-      throw new Error('model outputed symbolic tensor')
+      throw new Error('model output symbolic tensor')
     }
-    xs.dispose()
-
-    return { logits, ys }
-  }).mapAsync(async ({ logits, ys }) => {
-    const loss = (await tf.losses.softmaxCrossEntropy(ys, logits).array())
-    if (typeof loss !== 'number') {
+    const lossTensor = tf.losses.softmaxCrossEntropy(ys, logits)
+    const lossValue = await lossTensor.array()
+    if (typeof lossValue !== 'number') {
       throw new Error('got multiple loss')
     }
 
     const accTensor = tf.metrics.categoricalAccuracy(ys, logits)
     const accSize = accTensor.shape.reduce((l, r) => l * r, 1)
     const accSum = accTensor.sum()
-    const accSummed = await accSum.array()
+    const accSummed = accSum.arraySync()
```

Comment:
but why? async is highly concurrent when doing IO (which might happen depending on the running platform)

Reply:
Good catch, it was not intentional 👍